lib/BlockingFallbackResolver.php

```diff
@@ -2,42 +2,181 @@
+                if (\filter_var($name, FILTER_VALIDATE_IP, FILTER_FLAG_IPV4)) {
+                    return new Success([new Record($name, Record::A, null)]);
+                }
+
+                if (\filter_var($name, FILTER_VALIDATE_IP, FILTER_FLAG_IPV6)) {
+                    return new Failure(new DnsException("Got an IPv6 address, but type is restricted to IPv4"));
+                }
+                break;
+            case Record::AAAA:
+                if (\filter_var($name, FILTER_VALIDATE_IP, FILTER_FLAG_IPV6)) {
+                    return new Success([new Record($name, Record::AAAA, null)]);
+                }
+
+                if (\filter_var($name, FILTER_VALIDATE_IP, FILTER_FLAG_IPV4)) {
+                    return new Failure(new DnsException("Got an IPv4 address, but type is restricted to IPv6"));
+                }
+                break;
+            default:
+                if (\filter_var($name, FILTER_VALIDATE_IP, FILTER_FLAG_IPV4)) {
+                    return new Success([new Record($name, Record::A, null)]);
+                }
+
+                if (\filter_var($name, FILTER_VALIDATE_IP, FILTER_FLAG_IPV6)) {
+                    return new Success([new Record($name, Record::AAAA, null)]);
+                }
+                break;
+        }
+
+        $name = normalizeName($name);
+
+        // Follow RFC 6761 and never send queries for localhost to the caching DNS server
+        // Usually, these queries are already resolved via queryHosts()
+        if ($name === 'localhost') {
+            return new Success($typeRestriction === Record::AAAA
+                ? [new Record('::1', Record::AAAA, null)]
+                : [new Record('127.0.0.1', Record::A, null)]);
         }
 
-        return $this->query($name, Record::A);
+        return call(function () use ($name, $typeRestriction) {
+            if ($typeRestriction) {
+                return yield $this->query($name, $typeRestriction);
+            }
+
+            list(, $records) = yield Promise\some([
+                $this->query($name, Record::A),
+                $this->query($name, Record::AAAA),
+            ]);
+
+            return \array_merge(...$records);
+        });
     }
 
     public function query(string $name, int $type): Promise
     {
-        if ($type !== Record::A) {
-            return new Failure(new DnsException("Query for '{$name}' failed, because loading the system's DNS configuration failed and querying records other than A records isn't supported in blocking fallback mode."));
-        }
+        $name = $this->normalizeName($name, $type);
+        $question = $this->createQuestion($name, $type);
+
+        $request = $this->messageFactory->create(MessageTypes::QUERY);
+        $request->getQuestionRecords()->add($question);
 
-        $result = \gethostbynamel($name);
+        $encoder = $this->encoderFactory->create();
+        $question = $encoder->encode($request);
+
+        $result = \dns_get_record(...$question);
```

Comment:
Seems like a lot of unnecessary code and we can just call `\dns_get_record($this->normalizeName($name, $type), $type, null, null, true)`? Am I missing something?

Reply:
This is a nicer and more abstracted way of supporting multiple record types)